src/test/groovy/unit/org/pih/warehouse/requisition/RequisitionItemSpec.groovy

```diff
@@ -13,88 +16,346 @@ import spock.lang.Unroll
+        1        | 1                 || true        | RequisitionItemStatus.CANCELED
+        2        | 1                 || false       | RequisitionItemStatus.PENDING
+    }
+
+    void 'RequisitionItem.changeQuantity() should cancel old #quantity and assign #reasonCode and #comment'() {
+        given:
+        domain.product = new Product()
+        domain.quantity = quantity
+
+        when:
+        domain.changeQuantity(3, reasonCode.name(), comment)
+
+        then:
+        domain.quantityCanceled == quantity
+        domain.cancelReasonCode == reasonCode.name()
+        domain.cancelComments == comment
+
+        where:
+        quantity | reasonCode                       | comment
+        1        | ReasonCode.SUBSTITUTION          | "cancel comment"
+        5        | ReasonCode.REJECTED              | "cancel comment 2"
+        10       | ReasonCode.CANCELED_BY_REQUESTER | "cancel comment 3"
+    }
+
+    void 'RequisitionItem.changeQuantity() should reject applied #quantity and #reasonCode with error message including #fieldName'() {
+        given:
+        domain.product = new Product()
+        domain.quantity = 1
+
+        when:
+        domain.changeQuantity(quantity, reasonCode?.name(), "comment")
+
+        then:
+        final ValidationException exception = thrown()
+        exception.message.contains(fieldName)
+
+        where:
+        quantity | reasonCode              || fieldName
+        -1       | ReasonCode.SUBSTITUTION || "quantity"
+        null     | ReasonCode.SUBSTITUTION || "quantity"
+        2        | null                    || "cancelReasonCode"
+        0        | null                    || "cancelReasonCode"
+    }
+
+    void "RequisitionItem.approveQuantity() should calculate #quantityApproved when #quantity and #quantityCanceled are passed"() {
+        given:
+        domain.quantity = quantity
+        domain.quantityCanceled = quantityCanceled
+
+        when:
+        domain.approveQuantity()
+
+        then:
+        domain.quantityApproved == quantityApproved
+        notThrown(ValidationException)
 
         where:
-            quantity | quantityCancelled || isCancelled | status
-            1        | 1                 || true        | RequisitionItemStatus.CANCELED
-            2        | 1                 || false       | RequisitionItemStatus.PENDING
+        quantity | quantityCanceled || quantityApproved
+        2        | 1                || 1
+        10       | 5                || 5
+    }
+
+    void "RequisitionItem.approveQuantity() should throw validation exception when #quantity and #quantityCanceled are passed"() {
+        given:
+        domain.quantity = quantity
+        domain.quantityCanceled = quantityCanceled
+
+        when:
+        domain.approveQuantity()
+
+        then:
+        thrown(ValidationException)
+
+        where:
+        quantity | quantityCanceled
+        10       | 11
+        20       | 20
+        21       | 50
```

Comment:
what about quantityCancelled null and 0?

Reply:
Comparing null with value doesn't throw an exception, I can add this case in the test above